storage/txcache/txListForSender.go

```diff
@@ -0,0 +1,193 @@
+package txcache
+
+import (
+	"container/list"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/data"
+)
+
+// txListForSender is
+type txListForSender struct {
+	items          *list.List
+	mutex          sync.Mutex
+	copyBatchIndex *list.Element
+	copyBatchSize  int
+	orderNumber    int64
+	sender         string
+}
+
+// txListForSenderNode is a node of the linked list
+type txListForSenderNode struct {
+	txHash []byte
+	tx     data.TransactionHandler
+}
+
+// newTxListForSender creates a new (sorted) list of transactions
+func newTxListForSender(sender string, globalIndex int64) *txListForSender {
+	return &txListForSender{
+		items:       list.New(),
+		orderNumber: globalIndex,
+		sender:      sender,
+	}
+}
+
+// AddTx adds a transaction in sender's list
+// This is a "sorted" insert
+func (listForSender *txListForSender) AddTx(txHash []byte, tx data.TransactionHandler) {
+	// We don't allow concurent interceptor goroutines to mutate a given sender's list
+	listForSender.mutex.Lock()
+
+	nonce := tx.GetNonce()
+	mark := listForSender.findTxWithLargerNonce(nonce)
+	newNode := txListForSenderNode{txHash, tx}
+
+	if mark == nil {
+		listForSender.items.PushBack(newNode)
+	} else {
+		listForSender.items.InsertBefore(newNode, mark)
+	}
+
+	listForSender.mutex.Unlock()
+}
+
+func (listForSender *txListForSender) findTxWithLargerNonce(nonce uint64) *list.Element {
+	for element := listForSender.items.Front(); element != nil; element = element.Next() {
+		value := element.Value.(txListForSenderNode)
+		if value.tx.GetNonce() > nonce {
+			return element
+		}
+	}
+
+	return nil
+}
+
+// RemoveTx removes a transaction from the sender's list
+func (listForSender *txListForSender) RemoveTx(tx data.TransactionHandler) bool {
+	// We don't allow concurent interceptor goroutines to mutate a given sender's list
+	listForSender.mutex.Lock()
+
+	marker := listForSender.findTx(tx)
+	isFound := marker != nil
+	if isFound {
+		listForSender.items.Remove(marker)
+	}
+
+	listForSender.mutex.Unlock()
+
+	return isFound
+}
+
+// RemoveHighNonceTxs removes "count" transactions from the back of the list
+func (listForSender *txListForSender) RemoveHighNonceTxs(count uint32) [][]byte {
+	removedTxHashes := make([][]byte, count)
+
+	listForSender.mutex.Lock()
+
+	index := uint32(0)
+	var previous *list.Element
+	for element := listForSender.items.Back(); element != nil && count > index; element = previous {
+		// Remove node
+		previous = element.Prev()
+		listForSender.items.Remove(element)
+
+		// Keep track of removed transaction
+		value := element.Value.(txListForSenderNode)
+		removedTxHashes[index] = value.txHash
+
+		index++
+	}
+
+	listForSender.mutex.Unlock()
+
+	return removedTxHashes
+}
+
+func (listForSender *txListForSender) findTx(txToFind data.TransactionHandler) *list.Element {
+	for element := listForSender.items.Front(); element != nil; element = element.Next() {
+		value := element.Value.(txListForSenderNode)
+		if value.tx == txToFind {
```

Comment:
I think this will fail if the interface pointers are different even if the values are the same.

Reply:
It seems that it's ok if the interface pointers point to the same *transaction in memory (which is our case as far as I know, since we don't make copies).

Example: https://play.golang.org/p/t20Y9ZmIFpC

Anyway, now we use the nonce in `findTx()`.

Reply:
Right, this is just used internally by the txcache so calling findTx should be done only after getting the tx pointer from the hash to tx pointer map, which will hold the original pointer, not a copy. 

So I take back my comment